### ISSUE TITLE
Unclear CheckoutShippingAddressUpdate mutation needs refactoring

### ISSUE THREAD

**Issue by Kwaidan00** [mirumee/saleor#4493]
### What I'm trying to achieve

I was tried to add some additional logic when checkout's shipping address is changed.
Unfortunately, mutation `CheckoutShippingAddressUpdate` is unclear and brings me to the buggy solution.

It's the mutation: https://github.com/mirumee/saleor/blob/7a21736a3363aa8f1bacac7218c4a78e44e09652/saleor/graphql/checkout/mutations.py#L342-L376
When I read it, what steps I see:
1. get the checkout
2. validate new shipping address
3. clean shipping method
4. new shipping address is saved and assigned to the checkout.
The following steps are described in #4321 too. But, in fact, it does something different in some cases.

Method `validate_address` not only validates address, but it modifies given instance. If checkout had no shipping address - everything works great. But if checkout has shipping address and we would like to change it, it would be changed inside `validate_address` method. Function `change_shipping_address_in_checkout` will do nothing in that case. 

### Describe a proposed solution
I think that function should be refactored, it could be done by one of the following:
- change `validate_address` function name and docstring
- move saving valid address instance outside `validate_address` to make it clear

I think some additional tests should be written for the case that checkout has a shipping address and user wants change it - all tests I've found covers the case that `checkout.shipping_address` is `None` before sending mutation.

### Other solutions I've tried and won't work
…

**Comment by KevBelisle** [comment]
I'll take a look!

**Comment by KevBelisle** [comment]
Ok, so `validate_address` can be summarized as 3 steps:

- Validate `address_data` and end up with an `address_form`
- Create or update the `instance` argument with values from `address_form.cleaned_data`
- Return the updated `instance` or the `instance` created

I removed step 2 and instead had `validate_address` return the `address_form.cleaned_data`.

This means `perform_mutation` now looks like this:
```python
    @classmethod
    def perform_mutation(cls, _root, info, checkout_id, shipping_address):
        # Many unchanged lines of code skipped for brevity

        cleaned_address_data = cls.validate_address(shipping_address)

        checkout.shipping_address = cls.construct_instance(
            checkout.shipping_address or Address(),
            cleaned_address_data
        )
        cls.clean_instance(checkout.shipping_address)

        # Nothing changed below this comment

        update_checkout_shipping_method_if_invalid(checkout, info.context.discounts)

        with transaction.atomic():
            shipping_address.save()
            change_shipping_address_in_checkout(checkout, shipping_address)
        recalculate_checkout_discount(checkout, info.context.discounts)

        return CheckoutShippingAddressUpdate(checkout=checkout)
```


@Kwaidan00 Is this what you had in mind?

If so, there are many other places that use `validate_address` that will need to be refactored.

**Comment by Kwaidan00** [comment]
@KevBelisle thanks for taking a look at this issue :)
I think moving _create or update_ step outside validator is a good move. But there is still one thing that concerns me - the relation with `change_shipping_address_in_checkout` function. It's still weird to me that old `shipping_address` instance is updated every time and `change_shipping_address_in_checkout` do anything only for changing from `None` to first instance.
The Big Motivation for my issue was recognizing the situation address was changed. I thought I can add my code to the `change_shipping_address_in_checkout` function, but I could not, because address was changed earlier. In your proposal it's still the same problem.

**Comment by KevBelisle** [comment]
I'm not sure I understand what you mean... Are you saying you would prefer it if `checkout` was not modified until `change_shipping_address_in_checkout` is called?

If so, that makes sens. I modified the code in the Pull Request.

I also had to move the call to `update_checkout_shipping_method_if_invalid(checkout, info.context.discounts)` until after shipping_address is updated in `checkout` and saved. I'm not sure if this will cause problems.
